qiskit/circuit/_standard_gates_commutations.py

```diff
@@ -0,0 +1,3263 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Dictionary containing commutation relations of unparameterizable standard gates. The dictionary key
+is defined by a pair of gates and, optionally, their relative placement to each other, i.e.:
+
+   ┌───┐┌───┐
+q: ┤ X ├┤ Z ├    -> standard_gates_commutations["x", "z"] = False
+   └───┘└───┘
+
+or
+          ┌───┐
+q_0: ──■──┤ X ├
+     ┌─┴─┐└─┬─┘  -> standard_gates_commutations["cx", "cy"][1, 0] = False
+q_1: ┤ Y ├──■──
+     └───┘
+
+or
+
+q_0: ──■───────
+     ┌─┴─┐
+q_1: ┤ X ├──■──  -> standard_gates_commutations['cx', 'cx'][None, 0] = False
+     └───┘┌─┴─┐
+q_2: ─────┤ X ├
+          └───┘
+Note that the commutation of a pair of gates is only stored in one specific order, i.e. the commutation
+of a pair of cy and cx gates is only stored in standard_gates_commutations["cx", "cy"] and additionally
+as standard_gates_commutations["cx", "cy"]. The order of the gates is derived from the number of qubits
+in each gate and the value of the integer representation of the gate name.
+
+The values of standard_gates_commutations can be a Boolean value denoting whether the pair of gates
+commute or another dictionary if the relative placement of the pair of gates has an impact on the
+commutation. The relative placement is defined by the arrangement of the second gate's qubits with
+respect to the qubits of the first gate. In the second example, the zeroth qubits of cy overlaps with
+qubit 1 of cx. Furthermore, non-overlapping qubits are specified as 'None' in the dictionary key, e.g.
+example 3, there is no overlap on the zeroth qubit of the first cx but the zeroth qubit of the second
+gate overlaps with qubit 1 of the first cx.
```

Comment:
I am a bit confused about this. In the picture, the first gate is `CX(0, 1)` and the second gate is `CX(1, 2)`. Hence:
* qubit 0 of the first gate does not overlap with qubits of the second gate
* qubit 1 of the first gate overlaps with qubit 0 of the second gate

This is exactly what is written above and correlates to the key `standard_gates_commutations['cx', 'cx'][None, 0]`.
However, it also says:
> The relative placement is defined by the arrangement of the second gate's qubits with
respect to the qubits of the first gate. 

So, should this be the other way around: i.e. the arrangement of the first gate's qubits with respect to the second gate?

Reply:
I agree the definition was not concise enough, what do you think about 

- The relative placement is defined by the gate qubit arrangements as q2^{-1}[q1[i]] where q1[i] is the ith qubit of the first gate and q2^{-1}[q] returns the qubit index of qubit q in the second gate (possibly 'None')